# SQL Queries and AACT Data

In this assignment, you will answer questions about clinical trials registered with the US government. The data is a subset of the data made available through the Clinical Trials Transformation Initiative. 
The data provided is a subset of the full data set, so be sure to use the
version of the data available on the course Canvas site.  If you are
interested, you can learn more about the dataset at
http://aact.ctti-clinicaltrials.org.


## What's In and Out of Scope

This is intended to be a declarative SQL query assignment. Therefore, you must write
queries in SQL (not functions, and not in Python).  You
may use VIEWs as needed and you may use standard built-in PostgreSQL
functions (e.g. ROUND or CASE statements). If you're not sure if
something is allowed, ask!

## Getting Started

First,  go to your database, and create the tables found in the tablesA2.sql file.

## Load the data

Load the data needed for the assignment. You should do this in
pgAdmin or psql. The files are provided in the Canvas assignment. They are .sql files. You can copy & paste the contents into a pgAdmin window and run them or use the following command in psql:

\i <filename.sql>

Where <filename.sql> is the name of the file you want to run.  For example:

\i conditions.sql


## Cautions
You must use the table and attribute names provided. Do not rename anything. We want to be able to run your submission and see all of the  results be generated.


At the end of your submission, include **DROP** or **DELETE** statements that remove any VIEWs or TABLEs your code created.

## Turnin
Fill in this notebook and download it as a notebook (.ipynb).  By 11:55P on the due date, submit this document electronically to Canvas.   Answers in the comments section of CANVAS are not acceptable and will not be read.  Be sure to turn in the file with your answers in it, not just the empty assignment notebook


## Grading
 The number of points for each query is indicated in the question.   
If you don't get the right answer or your code is not correct, you won't get all of the points; partial credit may be given
at the discretion of the grader.

## Academic Honesty

The following level of collaboration is allowed on this assignment: You
may discuss the assignment with your classmates at a high level. Any
issues getting Postgres running is totally fine. What
is not allowed is direct examination of anyone else's SQL code (on a
computer, email, whiteboard, etc.) or allowing anyone else to see your
SQL code. You may not discuss query results with your classmates.

You may use the search engine of your choice to lookup the syntax for
SQL commands, but may not use it to find answers to queries.



## Connect to your database

Run the next cell to enter your netId 

In [1]:
import getpass
myNetId = input()


sy31


And the next cell to enter your password

In [2]:
password = getpass.getpass()


········


In [3]:
# set this value to 430 or 533, based on the course you are taking.
course = "533"

In [4]:
def make_conn_str(netId, password):
      
    hostname = "comp" + course + ".clear.rice.edu"
    db = netId + '_db'
    
    return f"postgresql://{netId}:{password}@{hostname}/{db}"

In [5]:
conn_str = make_conn_str(myNetId, password)

In [21]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [7]:
%sql $conn_str

'Connected: sy31@sy31_db'

In [22]:
%sql sqlite:///writers.db3

'Connected: @writers.db3'

Check connectivity by running the next cell. The query should return 2 records.

In [4]:
%%sql 
.table

 * sqlite:///writers.db3
(sqlite3.OperationalError) near ".": syntax error
[SQL: .table]
(Background on this error at: http://sqlalche.me/e/e3q8)


###  Create the database tables and load the data into the database

* Upload the files to your Jupyter envirnoment, using the upload button in the upper right
* Open a terminal window by selecting "New / Terminal" from the upper right

In the following steps, replace yournetidhere with your netId and XXX with 430 or 533, as appropriate.

* Run ```psql``` to connect to your database. 

```
psql sslmode=require -U sy31 -h comp533.clear.rice.edu -d sy31_db
```
* Once you are connected to the database use the following commands to create the tables and load the data.

    ```\i A2F19tabledefs.sql``` 
    
    ```\i responsible_parties.sql``` 
    
    ```\i overall_officials.sql``` 
    
    ```\i sponsors.sql``` 
    
    ```\i factor.sql``` 
         
    ```\i studies.sql``` 
    
    ```\i conditions.sql``` 
    

## Queries
Answer all of the questions below by writing and executing SQL queries. The queries must contain ONLY the answer to the question (no extra rows or columns). You may need to explore the database a bit prior to generating your final solutions. 

1 (8 points) We want a master list of the names of all the people in charge in the database.  Write queries to fill in the values in the following table  

|                     | # Names | # Unique Names |
|---------------------|---------|----------------|
| responsible_parties | 1       | 2              |
| overall_officials   | 3       | 4              |
| sponsors            | 5       | 6              |
| total               | 7       | 8              |


In [7]:
%%sql sqlite://
-- query for entry 1
select count(name) from responsible_parties;


Done.


[(112214,)]

In [6]:
\i A2F19tabledefs.sql

SyntaxError: unexpected character after line continuation character (<ipython-input-6-f75484ed3345>, line 1)

In [8]:
%%sql sqlite://
-- query for entry 2
select count(distinct(name)) from responsible_parties;


Done.


[(61334,)]

In [10]:
%%sql
-- query for entry 3
select count(name) from overall_officials;


 * sqlite:///writers.db3
Done.


[(0,)]

In [ ]:
%%sql
-- query for entry 4
select count(distinct(name)) from overall_officials;


In [ ]:
%%sql
-- query for entry 5
select count(name) from sponsors;


In [ ]:
%%sql
-- query for entry 6
select count(distinct(name)) from sponsors;


In [ ]:
%%sql
-- query for entry 7
select count(U.name) from (select name from sponsors  union all select name from overall_officials union all select name from responsible_parties) as U;


In [ ]:
%%sql
-- query for entry 8
select count(U.name) from (select name from sponsors  union  select name from overall_officials union  select name from responsible_parties) as U;


Replace the numbers in this table with your query results

|                     | # Names | # Unique Names |
|---------------------|---------|----------------|
| responsible_parties | 112214       | 61334              |
| overall_officials   | 276769       | 180010              |
| sponsors            | 411461       | 47480              |
| total               | 800444       | 288824              |

9 (2 points) Are there still duplicate entries in the table? If there are, modify query 8 to remove them. Include your query and the final count of the number of unique names across the 3 tables.



your text answer here
|                     | # Names | # Unique Names |
|---------------------|---------|----------------|
| responsible_parties | 112214       | 61334              |
| overall_officials   | 276769       | 180010              |
| sponsors            | 411461       | 47480              |
| total               | 800444       | 275386              |

In [ ]:
%%sql
select count(U.name) from (select distinct(name) from sponsors  union  select name from overall_officials union  select name from responsible_parties) as U;


10 (4 points) Sort the results of the previous query and examine the results. Give 2 different examples of names that are likely to be duplicates, but are showing up as separate entries. The reason why they are showing up as duplicates must be different for the two examples you provide. Provide the reason why you believe the names are the same.


your text answer here
A.A Voors
Adriaan A. Voors, MD, PhD;
A. Eden Evins
A. Eden Evins, M.D., M.P.H.
 
The Dr. Voors has duplicates because A.A is an abbrev for Adriaan A.;
Eden Evins has duplicates because one include the degree and the other doesn't. 
Each pair of duplicate has same affliations, so I believe they are duplicates.

11 (4 points) Make 2 suggestions for how to change the data storage structure to reduce the number of these remaining duplicates. 



Use same format for names; do not include degree in names.

Studies have shown that lifestyle choices can have a big impact on longevity. Let's investigate which studies deal with lifestyle aspects. In particular, we will look at conditions pertaining to 
* alcohol use
* diet
* lifestyle
* nutrition
* sexual activity
* substance abuse
* tobacco use
* weight

These factors are all in the ```factor``` table. This table contains two attributes: ```category``` and ```condition```. The conditions are drawn from the ```conditions table```, and are grouped into the categories mentioned above.

Let's compute a score that incorporates these factors for each study.

12 (2 points) How many studies have conditions that fall into the category of 'nutrition'?

In [9]:
%%sql sqlite://
select count( *) from factor where category='nutrition';

Done.


[(193,)]

13 (10 points) Compute a risk factor score that is just the total number of factors covered by the study, ignoring the factor category completely. List the nct_ids and counts of the factors for every study that has at least 4 factors. Sort by nct_id.

In [ ]:
%%sql
select nct_id, count( *) from conditions group by nct_id having  count(nct_id)>=4 ;

14 (10 points) Compute a category score  for each study based on the number of categories covered. Add 1 to the study score if the study has ANY condition listed in the factor table for each category.  Each category can be counted at most once.

How many studies have category scores of at least 3?

In [ ]:
%%sql sqlite://


15 (10 points) How many studies in our database cover none of the conditions in the factor table? 


In [13]:
%%sql sqlite://
select count(studies.nct_id) as 'not cover number' from studies where studies.nct_id not in (select nct_id from conditions)

Done.


[(51,)]

16 (10 points) Rewrite the same query as immediately above, but using a different SQL mechanism (e.g. JOINs vs.  subqueries)

In [ ]:
%%sql

17 (10 points)
What are the nct_ids of other studies that have all of the conditions in study NCT00507767? Note: you may not hard-code the condition names from study NCT00507767 in your query. If there is more than one study, order by nct_id.


In [ ]:
%%sql

18 For this question, consider ONLY the data from studies with start\_date of 2016-06-01, except where otherwise noted (e.g. specific nct_ids)

The Jaccard Index (\url{https://en.wikipedia.org/wiki/Jaccard_index}) provides a similarity measure over sets. We can use the conditions as the set elements for calculating the Jaccard index, taking into consideration conditions in common between two studies and the total number of conditions for each study.

Note: Be careful about integer division in your calculations. If you divide two integers, you will get another integer. In this case, we want a decimal value. Consider using the NUMERIC data type when calculating the Jaccard Index.


 Round all answers to 2 decimal places.

18a (10 points) What is the Jaccard Index for studies with nct\_id NCT03330301 and NCT02742597?
You may use the nct\_ids in your query(s).



In [ ]:
%%sql

18b (no points, you need this for the next part) Compute the Jaccard Index for every pair of studies, based on conditions. 

Depending on how you write your queries, they can take a very long time to run. So, you might give some thought as to how to make them more efficient.  Note that there is a base case for computing the Jaccard Index, where if the number of conditions you are comparing from both studies is 0, the Jaccard index is defined to be 1.



In [ ]:
%%sql

18c (15 points) What is the average non-zero Jaccard index value in our set of studies? In other words, if you exclude pairs for which the Jaccard index value is 0, what is the average score? 

In [ ]:
%%sql

18d (5 points)
What percentage of study pairs have a Jaccard Index of 1?

In [ ]:
%%sql

In [22]:
%%sql sqlite://
select round(
(select count(distinct(c.name)) from conditions c where c.nct_id='NCT03330301' and c.name in (select c1.name from conditions c1 where c1.nct_id='NCT02742597'))
* 1.0 /
(select count(distinct(c.name)) as v1 from conditions c where c.nct_id='NCT03330301' or c.nct_id='NCT02742597')  
 , 2)
as Jaccard
where
()

Done.


[(0.07,)]

In [18]:
%%sql sqlite://
select 
    distinct
    t1.category as n1,
    t2.category as n2
from 
    factor t1,
    factor t2
where n1 < n2

Done.


[('alcohol use', 'diet'),
 ('alcohol use', 'lifestyle'),
 ('alcohol use', 'nutrition'),
 ('alcohol use', 'sexual activity'),
 ('alcohol use', 'substance abuse'),
 ('alcohol use', 'tobacco use'),
 ('alcohol use', 'weight'),
 ('diet', 'lifestyle'),
 ('diet', 'nutrition'),
 ('diet', 'sexual activity'),
 ('diet', 'substance abuse'),
 ('diet', 'tobacco use'),
 ('diet', 'weight'),
 ('lifestyle', 'nutrition'),
 ('lifestyle', 'sexual activity'),
 ('lifestyle', 'substance abuse'),
 ('lifestyle', 'tobacco use'),
 ('lifestyle', 'weight'),
 ('nutrition', 'sexual activity'),
 ('nutrition', 'substance abuse'),
 ('nutrition', 'tobacco use'),
 ('nutrition', 'weight'),
 ('sexual activity', 'substance abuse'),
 ('sexual activity', 'tobacco use'),
 ('sexual activity', 'weight'),
 ('substance abuse', 'tobacco use'),
 ('substance abuse', 'weight'),
 ('tobacco use', 'weight')]

In [16]:
%%sql sqlite://
select count(distinct(category)) from factor;

Done.


[(8,)]

In [23]:
%%bash
ls -lh

total 366328
-rw-r--r--   1 zhuliu  staff    30K Sep 24 21:47 A2-SQL.ipynb
drwx------  10 zhuliu  staff   320B Sep 24 21:40 A2F19data
-r-xr-xr-x   1 zhuliu  staff    19M Sep 22 20:26 A2F19data.zip
-rw-r--r--   1 zhuliu  staff   2.1K Sep 22 21:42 Untitled.ipynb
-rw-r--r--   1 zhuliu  staff    79M Sep 22 20:49 data.db3
-rw-r--r--   1 zhuliu  staff    79M Sep 22 20:49 writers.db3
